### PR TITLE
[SPARK-44527][SQL] Replace ScalarSubquery with null if its maxRows is 0

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR enhances `ConstantFolding` to replace ·ScalarSubquery· with null if its maxRows is 0.

### Why are the changes needed?

Simplify expression to improve query performance. For example:
```scala
spark.sql("CREATE TABLE t1 using parquet AS SELECT * FROM range(10)")
spark.sql("CREATE TABLE t2 using parquet AS SELECT * FROM range(10)")

spark.sql("SELECT * FROM t1 WHERE t1.id = (SELECT id FROM t2 WHERE false)").explain(true)
```

Before this PR:
```
== Optimized Logical Plan ==
Filter (isnotnull(id#3L) AND (id#3L = scalar-subquery#2 []))
:  +- LocalRelation <empty>, [id#4L]
+- Relation spark_catalog.default.t1[id#3L] parquet
```

After this PR:
```
== Optimized Logical Plan ==
LocalRelation <empty>, [a#7L, b#8L]
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala [hunk 1]
```diff
@@ -48,10 +48,7 @@ abstract class PropagateEmptyRelationBase extends Rule[LogicalPlan] with CastSup
   // This tag is used to mark a repartition as a root repartition which is user-specified
   private[sql] val ROOT_REPARTITION = TreeNodeTag[Unit]("ROOT_REPARTITION")
 
-  protected def isEmpty(plan: LogicalPlan): Boolean = plan match {
-    case p: LocalRelation => p.data.isEmpty
-    case _ => false
-  }
+  protected def isEmpty(plan: LogicalPlan): Boolean = SimplifyBinaryComparison.isEmpty(plan)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala:51]: Reference SimplifyBinaryComparison here looks a little strange.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -529,6 +530,15 @@ object SimplifyBinaryComparison
         case TrueLiteral EqualNullSafe b if !b.nullable => b
         case a EqualNullSafe FalseLiteral if !a.nullable => Not(a)
         case FalseLiteral EqualNullSafe b if !b.nullable => Not(b)
+
+        case EqualNullSafe(a: ScalarSubquery, b: ScalarSubquery)
+            if a.plan.maxRows.contains(0) && b.plan.maxRows.contains(0) =>
+          TrueLiteral
+        case e: EqualNullSafe => e
+        case BinaryComparison(a: ScalarSubquery, _) if a.plan.maxRows.contains(0) =>
+          Literal(null, BooleanType)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:539]: Although all the current sub class of `BinaryComparison` are also with NullIntolerant.
In order to doesn't introduce unexpected bug, shall we check `isInstanceOf[NullIntolerant]`?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Should move it to `ConstantFolding `.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -88,6 +88,10 @@ object ConstantFolding extends Rule[LogicalPlan] {
           e
       }
 
+    // Replace ScalarSubquery with null if its maxRows is 0
+    case s: ScalarSubquery if s.plan.maxRows.contains(0) =>
+      Literal(null, s.dataType)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:93]: +1. Move the case here, looks more clear. @wangyum Thank you.

**Review by beliefer — APPROVED** [review]
LGTM.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wangyum and @beliefer .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 4]
```diff
@@ -88,6 +88,10 @@ object ConstantFolding extends Rule[LogicalPlan] {
           e
       }
 
+    // Replace ScalarSubquery with null if its maxRows is 0
+    case s: ScalarSubquery if s.plan.maxRows.contains(0) =>
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:92]: @jchen5 Do we need to consider `mayHaveCountBug`?

**Pull request closed without merging**